File: client/Contracts/Frontier.cs
namespace BibleAtlas.Client.Contracts;

// ============================================================================
// THE FRONTIER CONTRACT — spec §4e compiled (skeleton-first, owner-reviewed
// in-chat 2026-09-07: the change map, the matrix, and the enum-not-strings
// correction are all owner rulings).
//
// THE SMART-FRONTIER RULE (owner verbatim): "frontiers need to be smart.
// I.e., they display instances of types of nodes on the frontier iff both:
// 1) they exist for the focus node and 2) we have not opted out of
// displaying that node type on the frontier of the focus node."
//   display(section, focus) ⟺ exists(instances, focus)
//                           ∧ ¬optedOut(focus.Kind, sectionType)
// The FrontierMatrix below IS the opt-out registry (one line per cell —
// the one-line-knob acceptance bar). Existence is the runtime gate: no
// empty section ever renders (hidden-when-empty and hatch deliverability
// are instances of the same law). Both gates are conformance-tested, both
// directions (§4e: implementing-without-content and rendering-without-
// implementing each fail standing tests).
//
// Same discipline as every Contracts file: EXTEND-ONLY for implementers;
// member changes route through the controller.
// ============================================================================

/// <summary>THE focus-kind vocabulary (renamed from NodeKind per the
/// owner's reuse challenge — the GRAPH's NodeKind lives in the Rust types
/// crate (id.rs) and is a different vocabulary; graph-types/src/frontier.rs
/// is the cross-language authority this file mirrors, incl. the
/// FocusKind→graph-NodeKind bridge). THE vocabulary — one enum (owner: "use our node type
/// enums rather than strings"). The compiler now catches a typo'd kind in
/// any matrix cell; a misspelled string would have silently opted a kind
/// out.</summary>
public enum FocusKind
{
    Verse,
    Passage,
    Chapter,
    Book,
    Author,
    Place,
    Person,
    Event,
    Catechism,
    Year,
    TimeAndPlace,
    PolityDelta,
    CommentaryItem,
    ConcordUnit,
}

/// <summary>The ONE string↔enum mapping site. IExplorable.Kind is
/// stringly-typed today; every kind-string switch in the codebase migrates
/// through here over time. An unknown string throws loud (fail-loud law) —
/// never a guess, never a default.</summary>
public static class FocusKinds
{
    public static FocusKind Parse(string kind) => kind switch
    {
        "Verse" => FocusKind.Verse,
        "Passage" => FocusKind.Passage,
        "Chapter" => FocusKind.Chapter,
        "Book" => FocusKind.Book,
        "Author" => FocusKind.Author,
        "Place" => FocusKind.Place,
        "Person" => FocusKind.Person,
        "Event" => FocusKind.Event,
        "Catechism" => FocusKind.Catechism,
        "Year" => FocusKind.Year,
        "TimeAndPlace" => FocusKind.TimeAndPlace,
        "PolityDelta" => FocusKind.PolityDelta,
        "CommentaryItem" => FocusKind.CommentaryItem,
        "ConcordUnit" => FocusKind.ConcordUnit,
        _ => throw new NotSupportedException(
            $"Unknown node kind '{kind}' — a new IExplorable kind must be " +
            "added to FocusKind/FocusKinds.Parse (Contracts, controller-routed) " +
            "before it can traverse the frontier contract."),
    };
}

// ── CAPABILITY INTERFACES (§4e) ─────────────────────────────────────────
// One per non-core section family. A kind implements one IFF it can
// genuinely provide the content today — the presence IS the claim; an
// event acquires cross-references the day event-level xref data exists,
// by joining the matrix cell, never by an empty section.
public interface IHasCrossReferences { }
public interface IHasParallels { }
public interface IHasEventMembership { }
public interface IHasPassageMembership { }
public interface IHasPersons { }
public interface IHasCatechismSupport { }
/// <summary>EVT-3: the refs-list under prior/following (accounts as refs,
/// never verse contents, per the owner's chronological-traversal order).</summary>
public interface IHasChronology { }
/// <summary>EVT-3: explorable Time:/Place: header rows; Place exploration =
/// f(location, event time) with the map-open side effect (§5 hatch).</summary>
public interface IHasTimeAndPlace { }
/// <summary>EVT-3: "events are keys that map to sets of Biblical accounts"
/// (owner verbatim) — the accounts render as explorable passages.</summary>
public interface IHasAccounts { }

// ── THE OPT-OUT REGISTRY (the matrix as compiled data) ──────────────────
// Absence of a kind from a set = opted out for that kind. Editing a cell
// is a ONE-LINE change. Body-only sections (Place's four, Catechism's
// four, PolityDelta's three, Person, CommentaryItem, the Chapter card)
// are core focus-presentation, not matrix entries.
public static class FrontierMatrix
{
    public static readonly IReadOnlySet<FocusKind> CrossReferences =
        new HashSet<FocusKind> { FocusKind.Verse, FocusKind.Passage };
    // Event deliberately absent from CrossReferences — the owner's
    // calibration row ("we are not yet at the point of being able ot
    // provide lots of cross references for events").

    public static readonly IReadOnlySet<FocusKind> Parallels =
        new HashSet<FocusKind> { FocusKind.Verse, FocusKind.Passage };

    public static readonly IReadOnlySet<FocusKind> EventMembership =
        new HashSet<FocusKind> { FocusKind.Verse };

    public static readonly IReadOnlySet<FocusKind> PassageMembership =
        new HashSet<FocusKind> { FocusKind.Verse };

    public static readonly IReadOnlySet<FocusKind> Persons =
        new HashSet<FocusKind> { FocusKind.Verse, FocusKind.Passage };

    public static readonly IReadOnlySet<FocusKind> CatechismSupport =
        new HashSet<FocusKind> { FocusKind.Verse, FocusKind.Passage };

    public static readonly IReadOnlySet<FocusKind> Chronology =
        new HashSet<FocusKind> { FocusKind.Event };

    public static readonly IReadOnlySet<FocusKind> TimeAndPlace =
        new HashSet<FocusKind> { FocusKind.Event };

    public static readonly IReadOnlySet<FocusKind> Accounts =
        new HashSet<FocusKind> { FocusKind.Event };
}

File: graph-types/src/frontier.rs
//! THE FRONTIER CONTRACT — the cross-language authority (owner-ordered in
//! Rust, 2026-09-07: "write it in rust and make sure it compiles before
//! implementing"). The C# client mirrors this module; a parity test holds
//! the mirror to it (the AQC scenario-count parity precedent).
//!
//! THE SMART-FRONTIER RULE (owner verbatim, spec §4e): "frontiers need to
//! be smart. I.e., they display instances of types of nodes on the
//! frontier iff both: 1) they exist for the focus node and 2) we have not
//! opted out of displaying that node type on the frontier of the focus
//! node."
//!
//!   display(section, focus) ⟺ exists(instances, focus)
//!                            ∧ allows(focus.kind, section.capability)
//!
//! `allows` below IS the opt-out registry. Rust buys us what no set-based
//! encoding can: the match is EXHAUSTIVE over both enums — adding a new
//! `FocusKind` (or a new `Capability`) refuses to compile until every cell
//! of the new row/column is an explicit decision. A cell edit is a
//! one-line change (the owner's standing acceptance bar).

/// THE focus-kind vocabulary — every EXPLORABLE kind the frontier serves.
/// NOT the graph's own `crate::id::NodeKind` (owner challenge 2026-09-07:
/// "we should just be reusing stuff that exists" — the answer: the graph
/// vocabulary has no Verse/Passage/Chapter/Book, which are LEVELS of
/// TextUnit/Container per id.rs's own doc, and Year/TimeAndPlace/
/// PolityDelta are client-synthesized views; the two vocabularies are
/// genuinely distinct, and the honest reuse is the type-checked BRIDGE
/// below, not a merged enum and not a colliding name).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FocusKind {
    Verse,
    Passage,
    Chapter,
    Book,
    Author,
    Place,
    Person,
    Event,
    Catechism,
    Year,
    TimeAndPlace,
    PolityDelta,
    CommentaryItem,
    ConcordUnit,
}

impl FocusKind {
    /// The type-checked bridge to the graph's own vocabulary
    /// (`crate::id::NodeKind`) — reuse, made explicit. `None` = a
    /// client-synthesized view with no single graph node kind behind it.
    pub const fn graph_kind(self) -> Option<crate::id::NodeKind> {
        use crate::id::NodeKind as G;
        Some(match self {
            FocusKind::Verse | FocusKind::Passage | FocusKind::ConcordUnit => G::TextUnit,
            FocusKind::Chapter | FocusKind::Book | FocusKind::Author => G::Container,
            FocusKind::Place => G::Place,
            FocusKind::Person => G::Person,
            FocusKind::Event => G::Event,
            FocusKind::Catechism => G::CatechismItem,
            FocusKind::CommentaryItem => G::CommentaryItem,
            FocusKind::Year | FocusKind::TimeAndPlace | FocusKind::PolityDelta => return None,
        })
    }
}

/// The segregated capability families (spec §4e) — one per non-core
/// frontier section family. Core (chrome, hatches, back/trail, the focus
/// body) is NOT here: core is unconditional law, never a capability.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Capability {
    CrossReferences,
    Parallels,
    EventMembership,
    PassageMembership,
    Persons,
    CatechismSupport,
    /// EVT-3: the refs-list under prior/following (accounts as refs).
    Chronology,
    /// EVT-3: explorable Time:/Place: rows; place exploration =
    /// f(location, event time) with the map-open side effect.
    TimeAndPlace,
    /// EVT-3: "events are keys that map to sets of Biblical accounts."
    Accounts,
}

/// THE OPT-OUT REGISTRY. `false` = opted out for that kind. Exhaustive
/// over both enums by construction — the compiler enforces the whole
/// matrix. One line per row; a cell edit is a one-line diff.
pub const fn allows(kind: FocusKind, cap: Capability) -> bool {
    use Capability as C;
    use FocusKind as K;
    match (kind, cap) {
        // Verse — the richest frontier.
        (K::Verse, C::CrossReferences | C::Parallels | C::EventMembership
            | C::PassageMembership | C::Persons | C::CatechismSupport) => true,
        (K::Verse, C::Chronology | C::TimeAndPlace | C::Accounts) => false,

        // Passage — verse-family, minus the verse-only memberships.
        (K::Passage, C::CrossReferences | C::Parallels | C::Persons
            | C::CatechismSupport) => true,
        (K::Passage, C::EventMembership | C::PassageMembership
            | C::Chronology | C::TimeAndPlace | C::Accounts) => false,

        // Event — the owner's calibration row: NO cross references until
        // event-level xref data exists ("we are not yet at the point of
        // being able ot provide lots of cross references for events").
        (K::Event, C::Chronology | C::TimeAndPlace | C::Accounts) => true,
        (K::Event, C::CrossReferences | C::Parallels | C::EventMembership
            | C::PassageMembership | C::Persons | C::CatechismSupport) => false,

        // Body-only kinds: their sections are core focus-presentation;
        // every capability is opted out until real data earns a cell.
        (
            K::Chapter | K::Book | K::Author | K::Place | K::Person
            | K::Catechism | K::Year | K::TimeAndPlace | K::PolityDelta
            | K::CommentaryItem | K::ConcordUnit,
            _,
        ) => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The owner's calibration row, pinned: an event frontier does not
    /// implement cross references (until event-level data exists — the
    /// day it does, this test is UPDATED alongside the one-line cell).
    #[test]
    fn event_has_no_cross_references() {
        assert!(!allows(FocusKind::Event, Capability::CrossReferences));
    }

    /// EVT-3's three event capabilities, pinned.
    #[test]
    fn event_implements_chronology_time_and_place_accounts() {
        assert!(allows(FocusKind::Event, Capability::Chronology));
        assert!(allows(FocusKind::Event, Capability::TimeAndPlace));
        assert!(allows(FocusKind::Event, Capability::Accounts));
    }

    /// Verse always carries its six; never the event-family three.
    #[test]
    fn verse_row_matches_the_owner_reviewed_matrix() {
        use Capability as C;
        for cap in [C::CrossReferences, C::Parallels, C::EventMembership,
                    C::PassageMembership, C::Persons, C::CatechismSupport] {
            assert!(allows(FocusKind::Verse, cap));
        }
        for cap in [C::Chronology, C::TimeAndPlace, C::Accounts] {
            assert!(!allows(FocusKind::Verse, cap));
        }
    }
}
